### ISSUE TITLE
"Error opening PCM device not found" when used remotely

### ISSUE THREAD

**Issue by troffasky** [pothosware/SoapyFCDPP#14]
Can see the Funcube Dongle is detected from CubicSDR devices list.
When I try to start streaming from CubicSDR, SoapySDRServer dies with:
```
SoapyServerListener::accept([::ffff:192.168.1.77]:45034)
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM not found
Error opening PCM device not found
```
Console output from CubicSDR:
```
[INFO] setup stream
Stream setup failed, stream is null. std::exception
SDR Thread stream init error.
[ERROR] SoapyLogAcceptor::handlerLoop(tcp://192.168.1.3:55132) FAIL: SoapyRPCUnpacker::recv(header) FAIL: 

```

Server version: 0.5.2-unknown
```
# SoapySDRUtil --probe
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Probe device 
[ERROR] avahi_service_browser_new() failed: Bad state
[INFO] makeFCDPP
[INFO] listAntennas
[INFO] listFrequencies
[INFO] getFrequencyRange
[INFO] getBandwidth
[INFO] listFrequencies
[INFO] getFrequencyRange
[INFO] getFrequencyArgsInfo
[INFO] listSampleRates
[INFO] listBandwidths

----------------------------------------------------
-- Device identification
----------------------------------------------------
  driver=FCDPP
  hardware=FCD

----------------------------------------------------
-- Peripheral summary
----------------------------------------------------
  Channels: 1 Rx, 0 Tx
  Timestamps: NO
  Other Settings:
     * period [key=period, default=0, type=int, range=[96, 48000, 960]]

----------------------------------------------------
-- RX Channel 0
----------------------------------------------------
  Full-duplex: NO
  Supports AGC: NO
  Stream formats: CS16, CF32
  Native format: CS16 [full-scale=32767]
  Full gain range: [0, 97] dB
    LNA gain range: [-5, 30, 2.5] dB
    Mixer gain range: [4, 12, 8] dB
    IF gain range: [3, 57, 1] dB
  Full freq range: [0.15, 240], [420, 1900] MHz
    RF freq range: [0.15, 240], [420, 1900] MHz
  Sample rates: 0.096 MSps
```
libsoapdysdr 0.8.1

**Comment by troffasky** [comment]
Not just remote, doesn't work locally with rtl_433 either:
```
# rtl_433 -d ""
rtl_433 version 21.05 (2020-05-09) inputs file rtl_tcp RTL-SDR SoapySDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/home/alex/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
Registered 157 out of 186 device decoding protocols [ 1-4 8 11-12 15-17 19-23 25-26 29-36 38-60 63 67-71 73-100 102-105 108-116 119 121 124-128 130-149 151-161 163-168 170-175 177-186 ]
[ERROR] avahi_service_browser_new() failed: Bad state
[INFO] makeFCDPP
[INFO] setup stream
ALSA lib pcm.c:2660:(snd_pcm_open_noupdate) Unknown PCM not found
Error opening PCM device not found
```
libasound2 version is 1.2.5.1, FWIW.

**Comment by troffasky** [comment]
I think my problem here is that this driver is for the Funcube Dongle Pro Plus, and I merely have a Pro [I wish the names were a bit more distinct!]. 
I will close this issue on that basis, but I am sure someone can reopen this if my assumption is wrong.

**Comment by phlash** [comment]
Hi, sorry I missed this earlier - it's definitely a problem somewhere as the driver supports all versions of FCD (and I apologise for the names, those are 'marketing' choices, you can use the version numbers, 1.0, 1,1 [Pro, 96k] or 2.0 (Pro+, 192k).

From your logs, it seems the driver is unable to find the appropriate sound device in the kernel that relates to the HID device it has found (for your FCD v1.x): `Error opening PCM device not found`

Please re-run a test on the local machine where the device is, with Soapy logging turned up to TRACE:
```bash
% SOAPY_SDR_LOG_LEVEL=TRACE SoapySDRUtil --make=driver=fcdpp
```

This will provide more detail on the HID device identity, and may determine if we have a bug in the parsing of that to detect the Alsa device name, _here's what I get_:
```
phlash@bruce:~$ SOAPY_SDR_LOG_LEVEL=TRACE SoapySDRUtil --make=driver=fcdpp
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Make device driver=fcdpp
[TRACE] findFCDPP
[TRACE] Found device: 0001:0006:02, hw:1,0
[TRACE] findFCDPP=1 devices
[INFO] makeFCDPP
[DEBUG] SoapyFCDPP('0001:0006:02','hw:1,0',0,0)
  driver=FCDPP
  hardware=FCD
```

Please also check that your device is registering as an Alsa sound card by looking at the content of:
```bash
% cat /proc/asound/cards
```
this should contain an entry for the FCD _similar to mine_:
```
phlash@bruce:~$ cat /proc/asound/cards
 0 [PCH            ]: HDA-Intel - HDA Intel PCH
                      HDA Intel PCH at 0xa1518000 irq 146
 1 [V10            ]: USB-Audio - FUNcube Dongle V1.0
                      Hanlincrest Ltd. FUNcube Dongle V1.0 at usb-0000:00:14.0-4, full speed
```
if so, please look at the content of the following (USB path) file:
```bash
% cat /proc/asound/card<n>/usbbus     <--- replace <n> with the entry number from above, in my case '1'
```
this should contain a pair of numbers that locate the FCD on the USB bus, again, _here's mine_:
```
phlash@bruce:~$ cat /proc/asound/card1/usbbus 
001/006
```
finally, check that these numbers match with the output of `lsusb`, here's mine:
```
phlash@bruce:~$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 004: ID 06cb:00a2 Synaptics, Inc. Metallica MOH Touch Fingerprint Reader
Bus 001 Device 003: ID 13d3:56a6 IMC Networks Integrated Camera
Bus 001 Device 002: ID 8087:0025 Intel Corp. Wireless-AC 9260 Bluetooth Adapter
Bus 001 Device 006: ID 04d8:fb56 Microchip Technology, Inc. FUNcube Dongle V1.0  <----- here it is@001/006
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
Hope this helps...

**Comment by troffasky** [comment]
```
# SOAPY_SDR_LOG_LEVEL=TRACE SoapySDRUtil --make=driver=fcdpp
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Make device driver=fcdpp
[TRACE] findFCDPP
[TRACE] Found device: 1-1.4:1.2, not found
[TRACE] findFCDPP=1 devices
[INFO] makeFCDPP
[DEBUG] SoapyFCDPP('1-1.4:1.2','not found',0,0)
  driver=FCDPP
  hardware=FCD
  
```
At this point I'm wondering if the above is telling me that it can't find the FCD at all, so I unplugged it and re-ran:

```
# SOAPY_SDR_LOG_LEVEL=TRACE SoapySDRUtil --make=driver=fcdpp
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Make device driver=fcdpp
[TRACE] findFCDPP
[TRACE] findFCDPP=0 devices
[INFO] makeFCDPP
Segmentation fault
```
So being plugged in does make a difference, thankfully :)

```
# cat /proc/asound/cards
 1 [V10            ]: USB-Audio - FUNcube Dongle V1.0
                      Hanlincrest Ltd. FUNcube Dongle V1.0 at usb-0000:00:1a.0-1.4, full speed
# cat /proc/asound/card1/usbbus
001/004
# lsusb | grep cube
Bus 001 Device 004: ID 04d8:fb56 Microchip Technology, Inc. FUNcube Dongle V1.0
```
qthid-2.2 does find the device and can toggle it from bootloader to application mode [confirmed by watching USB events with journalctl -f], although the display is a bit garbled.

**Comment by phlash** [comment]
Thank you! This seems to be where the problem is:

```[TRACE] Found device: 1-1.4:1.2, not found```

that shows me that your HID device path `1-1.4:1.2` is completely different to mine and thus fails to parse. Yours appears to be a full USB device/port path, whereas mine is only bus/device ID. What versions of the following are you building/running with?
 * kernel
 * libhidapi-libusb0

then we should be able to reproduce and fix (after reading some manuals!)

**Comment by troffasky** [comment]
Package: libhidapi-libusb0
Version: 0.11.0-1
Package: linux-image-5.14.0-2-amd64
Version: 5.14.9-2

from Debian SID.

**Comment by phlash** [comment]
OK, I've found the change in libhidapi-libusb0:
https://github.com/libusb/hidapi/commit/03998f441847afc41279471a1c4ab6e7d797eb6b#diff-516ed9d234e81614e6fa65cfc81a50b8b119e8a0769b2e66a1c89166bac437f7
which changes the device path from mine (Debian bulllseye/stable -> 0.10.1) to yours (Debian SID -> 0.11.0).

I'll grab the SID build of libhidapi-libusb0 and work on a fix :smile:

**Comment by phlash** [comment]
Righto - my fork now contains a fix that works for me: https://github.com/phlash/SoapyFCDPP.git

Please give it a try - I'll open a PR back to here.

**Comment by troffasky** [comment]
Yes, that's looking better, locally:

```
# SOAPY_SDR_LOG_LEVEL=TRACE SoapySDRUtil --make=driver=fcdpp
######################################################
##     Soapy SDR -- the SDR abstraction library     ##
######################################################

Make device driver=fcdpp
[TRACE] findFCDPP
[TRACE] Found device: 1-1.4:1.2, hw:1,0
[TRACE] findFCDPP=1 devices
[INFO] makeFCDPP
[DEBUG] SoapyFCDPP('1-1.4:1.2','hw:1,0',0,0)
  driver=FCDPP
  hardware=FCD
```

I've not had much luck remotely, however. 
sdrpp segfaults as soon as you start streaming. CubicSDR works but is really choppy and logs a lot of this:
```
SDRThread::readStream(): 2. SoapySDR read failed with code: -1
SDRThread::readStream(): 3.1 iqDataOutQueue output queue is full, discard processing of the batch...
SSDRThread::readStream(): 2. SoapySDR read failed with code: -1
SDRThread::readStream(): 2. SoapySDR read failed with code: -1
SDRThread::readStream(): 3.1 iqDataOutQueue output queue is full, discard processing of the batch...
SSDRThread::readStream(): 2. SoapySDR read failed with code: -1

```
Pretty sure it's not a network issue as it's a whole 2Mbps and I have a 1G link to the server. This might be nothing to do with SoapyFCPP of course.

**Comment by phlash** [comment]
Ah good - I'll let @guruofquality have time to look at the PR, I usually miss something like tagging...

By default the SoapyFCDPP driver batches up data into 250msec chunks, which can look quite lumpy as most drivers are nearer 50msec. It does this to avoid thread thrashing on small remote heads (ie: my OrangePi Zero!). You can change the batch size with a driver argument `period=<n>` where `<n>` is the number of samples in each batch. On my x64 it's happy down to 1024 (~5msec/batch at 192kHz), have a play..

**Comment by troffasky** [comment]
Yes, that's the fix. The period field in CubicSDR is unlabelled and defaults to 24000. Tried 1024 and it's absolutely fine.

**Comment by phlash** [comment]
Change merged in #15
